cmd/opampsupervisor/supervisor/supervisor.go

```diff
@@ -406,24 +438,60 @@ func (s *Supervisor) startOpAMP() error {
 		},
 		Capabilities: s.Capabilities(),
 	}
-	err = s.opampClient.SetAgentDescription(s.agentDescription)
-	if err != nil {
+	if err = s.opampClient.SetAgentDescription(s.agentDescription); err != nil {
 		return err
 	}
 
-	err = s.opampClient.SetHealth(&protobufs.ComponentHealth{Healthy: false})
-	if err != nil {
+	if err = s.opampClient.SetHealth(&protobufs.ComponentHealth{Healthy: false}); err != nil {
 		return err
 	}
 
 	s.logger.Debug("Starting OpAMP client...")
+	if err = s.opampClient.Start(context.Background(), settings); err != nil {
+		return err
+	}
+	s.logger.Debug("OpAMP client started.")
 
-	err = s.opampClient.Start(context.Background(), settings)
+	s.logger.Debug("Starting OpAMP server...")
+	if err = s.startOpAMPServer(); err != nil {
+		return err
+	}
+	s.logger.Debug("OpAMP server started.")
+
+	return nil
+}
+
+func (s *Supervisor) startOpAMPServer() error {
+	s.opampServer = server.New(newLoggerFromZap(s.logger))
+
+	var err error
+	s.opampServerPort, err = s.findRandomPort()
 	if err != nil {
 		return err
 	}
 
-	s.logger.Debug("OpAMP Client started.")
+	err = s.opampServer.Start(newServerSettings(flattenedSettings{
+		endpoint:         fmt.Sprintf("localhost:%d", s.opampServerPort),
+		onConnectingFunc: func(_ *http.Request) {},
+		onMessageFunc: func(_ serverTypes.Connection, message *protobufs.AgentToServer) {
+			s.logger.Debug("Received message")
+			if message.AgentDescription != nil {
+				s.agentDescription = message.AgentDescription
+			}
+			if message.EffectiveConfig != nil {
+				s.logger.Debug("Setting confmap")
+				fmt.Println(string(message.EffectiveConfig.ConfigMap.ConfigMap[""].Body))
+				s.effectiveConfig.Store(string(message.EffectiveConfig.ConfigMap.ConfigMap[""].Body))
```

Comment:
Can we safely assume that the entire config is reported this way? Or should we iterate on the map and merge?

Reply:
This is a bit of an open question. I will open an issue in the spec repo to link to here and we can use that as a starting point for discussion.

Reply:
Spec issue: https://github.com/open-telemetry/opamp-spec/issues/184